src/session.rs

```diff
@@ -0,0 +1,202 @@
+                    continue 'event_loop;
+                }
+
+                let language_id = path_to_language_id(&extensions, &request.meta.buffile);
+                if language_id.is_none() {
+                    debug!(
+                        "Language server is not configured for extension `{}`",
+                        ext_as_str(&request.meta.buffile)
+                    );
+                    continue 'event_loop;
+                }
+                // is_none + unwrap pattern to reduce nesting again
+                // (is it a sign that block should be broken down into functions?)
+                let language_id = language_id.unwrap();
+
+                let root_path = find_project_root(&languages[&language_id].roots, &request.meta.buffile);
+
+                let route = Route {
+                    session: request.meta.session.clone(),
+                    language: language_id.clone(),
+                    root: root_path.clone(),
+                };
+
+                debug!("Routing editor request to {:?}", route);
+
+                if controllers.contains_key(&route) {
+                    let controller = controllers.get(&route).unwrap();
+                    controller.sender.send(request);
+                } else {
+                    // because Kakoune triggers BufClose after KakEnd
+                    // we don't want textDocument/didClose to spawn new controller
+                    if request.method == notification::DidCloseTextDocument::METHOD {
+                        continue 'event_loop;
+                    }
+                    let controller = spawn_controller(
+                        config.clone(),
+                        route.clone(),
+                        request,
+                        editor_tx.clone(),
+                    );
+                    controllers.insert(route, controller);
+                }
+            }
+        }
+    }
+    stop_session(&mut controllers);
+}
+
+/// Reap controllers associated with editor session.
+fn exit_editor_session(controllers: &mut Controllers, request: &EditorRequest) {
+    info!(
+        "Editor session `{}` closed, shutting down associated language servers",
+        request.meta.session
+    );
+    // have to clone keys as controllers are mutated inside loop
+    for route in controllers.keys().cloned().collect::<Vec<_>>() {
+        if route.session == request.meta.session {
+            // should be safe to unwrap because we are iterating controllers' keys
+            let controller = controllers.remove(&route).unwrap();
+            info!("Exit {} in project {}", route.language, route.root);
+            // to notify kak-lsp about editor session end we use the same `exit` notification as
+            // used in LSP spec to notify language server to exit, thus we can just clone request
+            // and pass it along
+            controller.sender.send(request.clone());
```

Comment:
To properly clean up this without timeouts, we need to drop `controller.sender` first, and then block on `controller.is_alive.recv()`. Now with the code written, I think that we probably should move `sender` out of `Controller`, so that they could be destroyed automatically?

Smth like this:


```rust
struct ControllerHandle {
    is_alive: Sender<Void>,
    thread: JoinHandle<()>,
}

let mut handles: Vec<ControllerHandle> = Vec::new();
let mut controllers: HashMap<Route, Sender<EditroRequest>> = HashMap::new();
'event_loop: loop {
    let mut dead_controller = None;
    select! {
        recv(handles.iter().map(|it| &it.is_alive), msg, from) => {
            assert!(msg.is_none());
            dead_controller = handles.iter().position(|it| it.is_alive == from)
        }
        recv(commmands_receiver, ...) => {
            let route = ...
            let (handle, sender) = spawn_controller(route);
            handles.push(handle);
            controllers.insert(route, sender);
         }
    }
    if let Some(idx) = dead_controller {
        handles.swap_remove(idx).thread.join()
    }
}

drop(controllers); // stopped controllers
for handle in handles {
    assert!(handle.is_alive().recv().is_none());
    handle.thread.join().
}

```

Reply:
Note that this has a nice bonus: it supports multiple exiting controllers for a single route.

That is, we can start a controller for route `foo`, add it to the controllers map, then stop it, remove from the map, and, while it is still in `handles`, add another controller with the `foo` route.

Reply:
Hm, nope, this won't work as is: when we remove a handle from `handles`, we need to remove the corresponding route as well, and my proposal does not allow to do that. Perhaps it's just best to keep the `sender` in Controller, but make it an `Option` to allow for clean-up.